File: tcpchat-client/internal/handler.go
package internal

import (
	"bufio"
	"encoding/json"
	"fmt"
	"io"
	"net"
	"os"
	"strings"
	"time"
)

func HandleServerMessages(conn net.Conn) {
	reader := bufio.NewReader(conn)
	for {
		data, err := reader.ReadBytes('\n')
		if err == io.EOF {
			fmt.Println("The server is down.")
			os.Exit(1)
		}
		if err != nil {
			fmt.Println("Error when reading message: ", err.Error())
			os.Exit(1)
		}

		fmt.Printf("\033[2K\r")

		msg := message{}
		err = json.Unmarshal(data, &msg)
		if err != nil {
			fmt.Println("Cannot unmarshal payload: ", err.Error())
			return
		}

		event, payload := msg.Event, msg.Payload

		switch event {
		case ERROR:
			fmt.Println("Error:", payload)
		case MESSAGE:
			fmt.Println(payload)
		}

		fmt.Print("> ")
	}
}

func HandleInput(conn net.Conn) {
	reader := bufio.NewReader(os.Stdin)
	for {
		fmt.Print("> ")
		input, _ := reader.ReadString('\n')
		input = strings.Trim(input, "\n")
		input = strings.TrimSpace(input)

		args := strings.Split(input, " ")

		if len(args) == 0 ||
			(len(args) == 1 && EVENT(args[0]) != GET_ROOMS) {
			fmt.Printf("\r")
			fmt.Println("Error: Wrong format.")
			continue
		}

		conn.SetWriteDeadline(time.Now().Add(2 * time.Second))
		payload, _ := json.Marshal(
			message{
				Event:   EVENT(args[0]),
				Payload: strings.Join(args[1:], " "),
			},
		)
		payload = append(payload, '\n')
		_, err := conn.Write(payload)
		if err != nil {
			fmt.Println(err.Error())
		}
	}
}

File: tcpchat-client/src/handler.rs
use std::{
    io::{BufRead, BufReader, Write},
    net::TcpStream,
    process::exit,
    str::FromStr,
};

use crate::message::{Event, Message, create_message};

pub fn handle_user_input(mut writer: TcpStream) {
    let mut buffer = String::new();

    loop {
        print!("> ");
        std::io::stdout().flush().unwrap_or_else(|_| {
            println!("Something wrong");
        });

        match std::io::stdin().read_line(&mut buffer) {
            Err(e) => {
                eprintln!("Error reading input: {}", e);
                exit(1);
            }
            Ok(_) => 'handler: {
                let input = buffer.trim_matches('\n').trim();

                let args: Vec<&str> = input.split_whitespace().collect();
                if args.is_empty() || (args.len() == 1 && args[0] != Event::GetRooms.to_string()) {
                    print!("\r");
                    std::io::stdout().flush().unwrap_or_else(|_| {
                        println!("Something wrong");
                    });
                    println!("Error: Wrong format.");
                    break 'handler;
                }

                let e = match Event::from_str(args[0]) {
                    Ok(e) => e,
                    Err(e) => {
                        println!("{e}");
                        break 'handler;
                    }
                };

                let m = create_message(e, args[1..].join(" "));

                writer
                    .write_all(m.as_bytes())
                    .expect("Cannot write to socket");
            }
        }

        buffer.clear();
    }
}

pub fn handle_messages_from_server(reader: TcpStream) {
    let mut reader = BufReader::new(reader);
    let mut buffer = String::new();

    loop {
        match reader.read_line(&mut buffer) {
            Err(e) => {
                eprintln!("Error when read message from server {e}");
                exit(1);
            }
            Ok(0) => {
                break;
            }
            Ok(_) => {
                let msg: Message = match serde_json::from_str(&buffer) {
                    Ok(msg) => msg,
                    Err(e) => {
                        eprintln!("Error when deserialize server message {e}");
                        exit(1);
                    }
                };

                buffer.clear();
                print!("\033[2K\r");
                std::io::stdout().flush().unwrap_or_else(|_| {
                    println!("Something wrong");
                });

                match msg.event {
                    Event::Error => {
                        println!("Error: {}", msg.payload);
                    }
                    Event::Message => {
                        println!("{}", msg.payload);
                    }
                    _ => {
                        eprintln!("Unknow server message");
                    }
                }

                print!("> ");
                std::io::stdout().flush().unwrap_or_else(|_| {
                    println!("Something wrong");
                });
            }
        }
    }
}
